pkg/ccl/storageccl/external_sst_reader.go

```diff
@@ -76,44 +90,73 @@ func ExternalSSTReader(
 				return nil, err
 			}
 		}
-		return storage.NewMemSSTIterator(content, false)
+		inMemorySSTs = append(inMemorySSTs, content)
 	}
+	return storage.NewPebbleMultiMemSSTIterator(inMemorySSTs, false, iterOps)
+}
 
-	raw := &sstReader{
-		ctx:  ctx,
-		sz:   sizeStat(sz),
-		body: f,
-		openAt: func(offset int64) (ioctx.ReadCloserCtx, error) {
-			reader, _, err := e.ReadFileAt(ctx, basename, offset)
-			return reader, err
-		},
+// ExternalSSTReader returns a PebbleSSTIterator for the SSTs in external storage,
+// optionally decrypting with the supplied parameters.
+//
+// ctx is captured and used throughout the life of the returned iterator, until
+// the iterator's Close() method is called.
+func ExternalSSTReader(
+	ctx context.Context,
+	e []cloud.ExternalStorage,
+	filenames []string,
+	encryption *roachpb.FileEncryptionOptions,
+	iterOps storage.IterOptions,
+) (storage.SimpleMVCCIterator, error) {
+	if len(e) != len(filenames) {
+		return nil, errors.New("must supply an external store for each file")
+	}
+	if !remoteSSTs.Get(&e[0].Settings().SV) {
+		return newMemPebbleSSTReader(ctx, e, filenames, encryption, iterOps)
 	}
+	remoteCacheSize := remoteSSTSuffixCacheSize.Get(&e[0].Settings().SV)
+	readers := make([]sstable.ReadableFile, 0, len(filenames))
 
-	var reader sstable.ReadableFile = raw
+	for i, basename := range filenames {
+		// prevent capturing the loop variables by reference when defining openAt below.
+		basename := basename
```

Comment:
@jbowens if I comment these local vars out, the unit test I wrote in the PR still passes :(. You mentioned that the roachtest failures occur at block boundaries. I assume this occurs when pebbleIterator.Iter.ExternalFiles[i].ReadAt is called. Do you have any tips at how I could construct the SSTs in the unit test to ensure the test hits block boundaries? Maybe this is overkill, since we no know the bug is squashed and the test still adds good coverage.

Reply:
Does it trigger if you vary the number of keys per-file, eg, 10, 100, 1000?

Reply:
nope. I also changed the sst's to partially overlap, and still no repro. They currently look like this: 
```
// t3               a500--------------------a10000
//
// t2      a50--------------a1000
//
// t1   a0----a100
```